qiime2/sdk/usage.py

```diff
@@ -0,0 +1,359 @@
+               outputs: UsageOutputNames):
+
+        if not isinstance(action, UsageAction):
+            raise TypeError('Must provide an instance of UsageAction.')
+        action.validate(inputs, outputs)
+
+        _, action_signature = action.get_action()
+
+        input_opts = inputs.build_opts(action_signature, self._scope)
+        output_opts = outputs.build_opts(action_signature, self._scope)
+
+        derived_outputs = self._action_(action, input_opts, output_opts)
+        self._add_outputs_to_scope(outputs, derived_outputs)
+
+    def _action_(self, action: UsageAction,
+                 input_opts: dict, output_opts: list):
+        raise NotImplementedError
+
+    def _assert_has_line_matching_(self, ref, label, path, expression):
+        raise NotImplementedError
+
+    def get_result(self, ref):
+        return self._get_record(ref)
+
+    def _add_outputs_to_scope(self, outputs, derived_outputs):
+        outputs.validate_derived(derived_outputs)
+        for ref, result in derived_outputs.items():
+            self._push_record(ref, result)
+
+    def _push_record(self, ref, value):
+        return self._scope.push_record(
+            ref=ref, value=value,
+            assert_has_line_matching=self._assert_has_line_matching_)
+
+    def _get_record(self, ref):
+        return self._scope.get_record(ref)
+
+    def _get_records(self):
+        return self._scope.records
+
+
+class DiagnosticUsage(Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+
+    def _init_data_(self, ref, factory):
+        self._recorder.append({
+            'type': 'init_data',
+            'ref': ref,
+        })
+        return ref
+
+    def _merge_metadata_(self, ref, records):
+        self._recorder.append({
+            'type': 'merge_metadata',
+            'ref': ref,
+            'records_refs': [r.ref for r in records],
+        })
+        return ref
+
+    def _get_metadata_column_(self, ref, record, column_name):
+        self._recorder.append({
+            'type': 'get_metadata_column',
+            'ref': ref,
+            'record_ref': record.ref,
+            'column_name': column_name,
+        })
+        return ref
+
+    def _comment_(self, text):
+        self._recorder.append({
+            'type': 'comment',
+            'text': text,
+        })
+
+    def _action_(self, action, input_opts, output_opts):
+        self._recorder.append({
+            'type': 'action',
+            'action': action,
+            'input_opts': input_opts,
+            'output_opts': output_opts,
+        })
+        return output_opts
+
+    def _assert_has_line_matching_(self, ref, label, path, expression):
+        self._recorder.append({
+            'type': 'assert_has_line_matching',
+            'ref': ref,
+            'label': label,
+            'path': path,
+            'expression': expression,
+        })
+
+
+class ExecutionUsage(Usage):
```

Comment:
This one is also really straight-forward. I think this iteration is working really well!